source/simulator/solver.cc

```diff
@@ -253,25 +253,92 @@ namespace aspect
       S_block_tolerance(S_block_tolerance)
     {}
 
-    template <class PreconditionerA, class PreconditionerMp>
+    template <class PreconditionerA, class SInvOperator>
     unsigned int
-    BlockSchurPreconditioner<PreconditionerA, PreconditionerMp>::
+    BlockSchurPreconditioner<PreconditionerA, SInvOperator>::
     n_iterations_A() const
     {
       return n_iterations_A_;
     }
 
-    template <class PreconditionerA, class PreconditionerMp>
+    template <class PreconditionerA, class SInvOperator>
     unsigned int
-    BlockSchurPreconditioner<PreconditionerA, PreconditionerMp>::
+    BlockSchurPreconditioner<PreconditionerA, SInvOperator>::
     n_iterations_S() const
     {
-      return n_iterations_S_;
+      return S_inv_operator.n_iterations;
     }
 
-    template <class PreconditionerA, class PreconditionerMp>
+    template <class PreconditionerMp>
+
+    //Used to implement right preconditioner in which
+    //the schur complement is approximated by
+    //the pressure mass matrix weigted by the inverse of viscosity
```

Comment:
End it like a proper sentence:
```suggestion
    //the pressure mass matrix weighted by the inverse of viscosity.
```